dashboard/peers.go

```diff
@@ -0,0 +1,576 @@
+	"github.com/ethereum/go-ethereum/p2p"
+)
+
+const (
+	eventBufferLimit = 128 // Maximum number of buffered peer events.
+	knownPeerLimit   = 100 // Maximum number of stored peers, which successfully made the handshake.
+	attemptLimit     = 100 // Maximum number of stored peers, which failed to make the handshake.
+
+	// eventLimit is the maximum number of the dashboard's custom peer events,
+	// that are collected between two metering period and sent to the clients
+	// as one message.
+	// TODO (kurkomisi): Limit the number of events.
+	eventLimit = knownPeerLimit << 2
+)
+
+// peerContainer contains information about the node's peers. This data structure
+// maintains the metered peer data based on the different behaviours of the peers.
+//
+// Every peer has an IP address, and the peers that manage to make the handshake
+// (known peers) have node IDs too. There can appear more peers with the same IP,
+// therefore the peer container data structure is a tree consisting of a map of
+// maps, where the first key groups the peers by IP, while the second one groups
+// them by the node ID. The known peers can be active if their connection is still
+// open, or inactive otherwise. The peers failing before the handshake (unknown
+// peers) only have IP addresses, so their connection attempts are stored as part
+// of the value of the outer map.
+//
+// Another criteria is to limit the number of metered peers so that
+// they don't fill the memory. The selection order is based on the
+// peers activity: the peers that are inactive for the longest time
+// are thrown first. For the selection a fifo list is used which is
+// linked to the bottom of the peer tree in a way that every activity
+// of the peer pushes the peer to the end of the list, so the inactive
+// ones come to the front. When a peer has some activity, it is removed
+// from and reinserted into the list. When the length of the list reaches
+// the limit, the first element is removed from the list, as well as from
+// the tree.
+//
+// The active peers have priority over the inactive ones, therefore
+// they have their own list. The separation makes it sure that the
+// inactive peers are always removed before the active ones.
+//
+// The peers that don't manage to make handshake are not inserted into the list,
+// only their connection attempts are appended to the array belonging to their IP.
+// In order to keep the fifo principle, a super array contains the order of the
+// attempts, and when the overall count reaches the limit, the earliest attempt is
+// removed from the beginning of its array.
+//
+// This data structure makes it possible to marshal the peer
+// history simply by passing it to the JSON marshaler.
+type peerContainer struct {
+	// Bundles is the outer map using the peer's IP address as key.
+	Bundles map[string]*peerBundle `json:"bundles,omitempty"`
+
+	// activePeers contains the peers with opened connection in random order.
+	activePeers *list.List
```

Comment:
Could we use slices instead? Perhaps lets investigate either here or in a followup pr.